### ISSUE TITLE
$client->listDir throws Exception when there is no file in remote directory

### ISSUE THREAD

**Issue by terrylinooo** [lazzard/php-ftp-client#17]
```
Lazzard\FtpClient\Exception\FtpClientException: [FtpClientException] - Failed to get files list.
```
I found that there would be at least one file in that directory will work.

Expected:

Return empty array instead.

**Comment by AmraniCh** [comment]
Hi, Thanks for opening this issue.

Actually, this exception is thrown when calling the native function `ftp_nlst ``and it is returned false, the exception message that you've got is means that the native function just fails without raising any error, because if the error raised the library can catch this error message and throws exceptions with them.

```php
public function listDir($directory, $filter = self::FILE_DIR_TYPE, $ignoreDots = true)
{
    if (!$files = $this->wrapper->nlist($directory)) {
        throw new FtpClientException($this->wrapper->getErrorMessage()
            ?: "Failed to get files list.");
    }
}
```

You have to know that FTP servers can haves different implementation of the FTP protocol specification, therefore getting an error should not necessary means you code is wrong, it just a "compatibility issues".

I've tested this method with my own FTP account and the directory was empty, and it works it returns an empty array.

If you want to debug your FTP server and know the exact FTP reply when sending the `nlst` command, you can use a tool that I've built [FTP Bridge](https://github.com/lazzard/ftp-bridge).

**Comment by betsmoraes** [comment]
Hi, im facing same issue.
What make if the FTP server dont support ftp_nlst ?

**Comment by AmraniCh** [comment]
@betsmoraes try this code and tell me if you get the same issue :

```php
$conn = @ftp_connect('ftp.byethost7.com');

if (!is_resource($conn)) {
    die('Connection failed to the server.');
}

$login = @ftp_login($conn, 'b7_27666328', '0659630023');

if (!$login) {
    die('Logging failed to the server.');
}

ftp_pasv($conn, true);

print_r(@ftp_nlist($conn, 'htdocs/lol'));

```

**Comment by AmraniCh** [comment]
@betsmoraes Please try running this code and tell me if you get the same error :

```php
$conn = @ftp_connect('host');

if (!is_resource($conn)) {
    die('Connection failed to the server.');
}

$login = @ftp_login($conn, 'username', 'password');

if (!$login) {
    die('Logging failed to the server.');
}

ftp_pasv($conn, true);

print_r(@ftp_nlist($conn, $dir));

```

**Comment by AmraniCh** [comment]
$conn = @ftp_connect('ftp.byethost7.com');

if (!is_resource($conn)) {
    die('Connection failed to the server.');
}

$login = @ftp_login($conn, 'b7_27666328', '0659630023');

if (!$login) {
    die('Logging failed to the server.');
}

ftp_pasv($conn, true);

print_r(@ftp_nlist($conn, 'htdocs/lol'));

**Comment by AmraniCh** [comment]
I think the problem is some FTP servers do not send the DOTS files ('.', '..') when listing the files, so if a directory is empty then the `ftp_nlst` will return en empty array. The solution is to check for the false value and not with `!` operator : 

```php
if (($files = $this->wrapper->nlist($directory)) === false) {
    throw new FtpClientException($this->wrapper->getErrorMessage()
        ?: "Failed to get files list.");
}
```

**Comment by AmraniCh** [comment]
I think the problem is some FTP servers do not send the DOTS files ('.', '..') when listing the files, so if a directory is empty then the `ftp_nlst` will return en empty array. The solution is to check for the false value with `!==` and not with `!` operator : 

```php
if (($files = $this->wrapper->nlist($directory)) !== false) {
    throw new FtpClientException($this->wrapper->getErrorMessage()
        ?: "Failed to get files list.");
}
```

**Comment by betsmoraes** [comment]
Hi Amrani, follow:
> @betsmoraes Please try with this code and tell me if you get the same error :
> 
> ```
> $conn = @ftp_connect($host);
> 
> if (!is_resource($conn)) {
>     die('Connection failed to the server.');
> }
> 
> $login = @ftp_login($conn, $username, $password);
> 
> if (!$login) {
>     die('Logging failed to the server.');
> }
> 
> ftp_pasv($conn, true);
> 
> print_r(@ftp_nlist($conn, $dir));
> ```

Returns connection successfull:
Array ( [0] => media [1] => ondemand [2] => playlist.txt [3] => sounds )

> I think the problem is some FTP servers do not send the DOTS files ('.', '..') when listing the files, so if a directory is empty then the `ftp_nlst` will return en empty array. The solution is to check for the false value with `!==` and not with `!` operator :
> 
> ```
> if (($files = $this->wrapper->nlist($directory)) !== false) {
>     throw new FtpClientException($this->wrapper->getErrorMessage()
>         ?: "Failed to get files list.");
> }
> ```

Changing to this, return same error:  [FtpClientException] - Failed to get files list.
This changes makes fail another classes too. with same error.

Thank you!

**Comment by AmraniCh** [comment]
Fixed in #18. A new release will published soon. Thanks for your contrubution @betsmoraes @terrylinooo

**Comment by AmraniCh** [comment]
Version [1.4.1](https://github.com/lazzard/php-ftp-client/releases/tag/v1.4.1) is just released.
